src/client/mesh.cpp

```diff
@@ -222,161 +237,79 @@ void setMeshColorByNormalXYZ(scene::IMesh *mesh,
 		const video::SColor &colorY,
 		const video::SColor &colorZ)
 {
-	if (mesh == NULL)
+	if (!mesh)
 		return;
-
-	u16 mc = mesh->getMeshBufferCount();
-	for (u16 j = 0; j < mc; j++) {
-		scene::IMeshBuffer *buf = mesh->getMeshBuffer(j);
-		const u32 stride = getVertexPitchFromType(buf->getVertexType());
-		u32 vertex_count = buf->getVertexCount();
-		u8 *vertices = (u8 *)buf->getVertices();
-		for (u32 i = 0; i < vertex_count; i++) {
-			video::S3DVertex *vertex = (video::S3DVertex *)(vertices + i * stride);
-			f32 x = fabs(vertex->Normal.X);
-			f32 y = fabs(vertex->Normal.Y);
-			f32 z = fabs(vertex->Normal.Z);
-			if (x >= y && x >= z)
-				vertex->Color = colorX;
-			else if (y >= z)
-				vertex->Color = colorY;
-			else
-				vertex->Color = colorZ;
-		}
-	}
+	auto colorizator = [=] (video::S3DVertex *vertex) {
+		f32 x = fabs(vertex->Normal.X);
+		f32 y = fabs(vertex->Normal.Y);
+		f32 z = fabs(vertex->Normal.Z);
+		if (x >= y && x >= z)
+			vertex->Color = colorX;
+		else if (y >= z)
+			vertex->Color = colorY;
+		else
+			vertex->Color = colorZ;
+	};
+	applyToMesh(mesh, colorizator);
 }
 
 void setMeshColorByNormal(scene::IMesh *mesh, const v3f &normal,
 		const video::SColor &color)
 {
 	if (!mesh)
 		return;
+	auto colorizator = [normal, color] (video::S3DVertex *vertex) {
+		if (vertex->Normal == normal)
+			vertex->Color = color;
+	};
+	applyToMesh(mesh, colorizator);
+}
 
-	u16 mc = mesh->getMeshBufferCount();
-	for (u16 j = 0; j < mc; j++) {
-		scene::IMeshBuffer *buf = mesh->getMeshBuffer(j);
-		const u32 stride = getVertexPitchFromType(buf->getVertexType());
-		u32 vertex_count = buf->getVertexCount();
-		u8 *vertices = (u8 *)buf->getVertices();
-		for (u32 i = 0; i < vertex_count; i++) {
-			video::S3DVertex *vertex = (video::S3DVertex *)(vertices + i * stride);
-			if (normal == vertex->Normal) {
-				vertex->Color = color;
-			}
-		}
-	}
+template <float v3f::*U, float v3f::*V>
+static void rotateMesh(scene::IMesh *mesh, float degrees)
+{
+	degrees *= M_PI / 180.0f;
+	float c = std::cos(degrees);
+	float s = std::sin(degrees);
+	auto rotator = [c, s] (video::S3DVertex *vertex) {
```

Comment:
why not just capture degrees ?

Reply:
That’s the whole point: trigonometric functions are really heavy, much much heavier than even division.